### ISSUE TITLE
convert to hatch and src layout?

### ISSUE THREAD

**Issue by tlambert03** [napari/npe2#280]
this issue https://github.com/pypa/setuptools/issues/3557 is really making cross-package editable installs difficult... meaning it's hard to run `pip install -e npe2` in a napari dev environment and have it work... you get something like this:
```
In [1]: import npe2

In [2]: npe2
Out[2]: <module 'npe2' (<_frozen_importlib_external._NamespaceLoader object at 0x10cfb3f40>)>
```

supposedly, moving to a src layout fixes that for setuptools.   but hatch just gets this right regardless, and I'd love to convert this repo over to hatch.  any objections @DragaDoncila @nclack @jni @Czaki ?

**Comment by Czaki** [comment]
I do not have big experience with hatch. Especially with the short trial, I do not be able to enforce running tests via hatch. 

But independently of the select build backend, I support switching to `src` layout.

In the context of hatch I prefer to either switch to more than one project or keep everything on setuptools as it is better known by core-devs. But maybe I'm wrong and more core-devs are familiar with hatch

**Comment by Czaki** [comment]
> I do not be able to enforce running tests via hatch." ... but no need to push it if you're uncomfortable with it.

In theory, hatch could replace `tox` as it supports creating venv and running tests in it. But I was not able to get this working (also do not spend to much time on it). 


> I'm curious though, have you, however, tried to install package A (that uses setuptools) in editable mode and then tried to use package A from local package B? I

I do not remember such case but I may have outdated setuptools in part of the environments.

**Comment by tlambert03** [comment]
> In theory, hatch could replace tox as it supports creating venv and running tests in it.

yeah, but I'm not talking about that here (besides, we don't even use tox in npe2).  I was just talking about using `hatchling` as a build backend

**Comment by tlambert03** [comment]
anyway, see #281  for the src layout ... would be trivial to move to a hatchling build backend from there

**Comment by nclack** [comment]
It seems like converting to a src layout is the important thing. 💕 src layout.

Hatch looks like a tool that makes that easier to setup and maintain.  I like that it seems to help w version updates/publishing workflow.  I don't have any experience with it, but it looks good. Seems worth a try.

**Comment by tlambert03** [comment]
to be clear, all this really entails is changing the `build-system` table in pyproject to 

```toml
[build-system]
requires = ["hatchling", "hatch-vcs"]
build-backend = "hatchling.build"

[tool.hatch.version]
source = "vcs"
```

**Comment by Czaki** [comment]
> to be clear, all this really entails is changing the `build-system` table in pyproject to

ok, so such a basic change without using all hatch features is easy to manage.
